File: scripts/patchers/kernel_jb_patch_vm_fault.py
"""Mixin: KernelJBPatchVmFaultMixin."""

from .kernel_jb_base import NOP


class KernelJBPatchVmFaultMixin:
    def patch_vm_fault_enter_prepare(self):
        """NOP a PMAP check in _vm_fault_enter_prepare.
        Strict mode:
        - Resolve vm_fault_enter_prepare function via symbol/string anchor.
        - In-function only (no global fallback scan).
        - Require a unique BL site with post-call flag test shape.
        """
        self._log("\n[JB] _vm_fault_enter_prepare: NOP")

        # Try symbol first
        foff = self._resolve_symbol("_vm_fault_enter_prepare")
        if foff >= 0:
            func_end = self._find_func_end(foff, 0x2000)
            result = self._find_bl_tbz_pmap(foff, func_end)
            if result:
                self.emit(result, NOP, "NOP [_vm_fault_enter_prepare]")
                return True

        # String anchor: all refs to "vm_fault_enter_prepare"
        str_off = self.find_string(b"vm_fault_enter_prepare")
        candidate_sites = set()
        if str_off >= 0:
            refs = self.find_string_refs(str_off, *self.kern_text)
            funcs = sorted(
                {
                    self.find_function_start(adrp_off)
                    for adrp_off, _, _ in refs
                    if self.find_function_start(adrp_off) >= 0
                }
            )
            for func_start in funcs:
                func_end = self._find_func_end(func_start, 0x4000)
                result = self._find_bl_tbz_pmap(func_start, func_end)
                if result is not None:
                    candidate_sites.add(result)

        if len(candidate_sites) == 1:
            result = next(iter(candidate_sites))
            self.emit(result, NOP, "NOP [_vm_fault_enter_prepare]")
            return True
        if len(candidate_sites) > 1:
            self._log(
                "  [-] ambiguous vm_fault_enter_prepare candidates: "
                + ", ".join(f"0x{x:X}" for x in sorted(candidate_sites))
            )
            return False

        self._log("  [-] patch site not found")
        return False

    def _find_bl_tbz_pmap(self, start, end):
        """Find strict BL site used by vm_fault_enter_prepare guard path.

        Expected local shape:
          BL target(rare)
          LDRB wN, [xM, #0x2c]
          ... TBZ/TBNZ wN, #bit, <forward>
        Returns BL offset when the match is unique inside [start, end).
        """
        hits = []
        scan_start = max(start + 0x80, start)
        for off in range(scan_start, end - 0x10, 4):
            d0 = self._disas_at(off)
            if not d0 or d0[0].mnemonic != "bl":
                continue
            bl_target = d0[0].operands[0].imm
            n_callers = len(self.bl_callers.get(bl_target, []))
            if n_callers >= 20:
                continue

            d1 = self._disas_at(off + 4)
            if not d1 or d1[0].mnemonic != "ldrb":
                continue
            op1 = d1[0].op_str
            if "#0x2c" not in op1 or not op1.startswith("w"):
                continue

            reg = op1.split(",", 1)[0].strip()
            matched = False
            for delta in (8, 12, 16):
                d2 = self._disas_at(off + delta)
                if not d2:
                    continue
                i2 = d2[0]
                if i2.mnemonic not in ("tbz", "tbnz"):
                    continue
                if not i2.op_str.startswith(f"{reg},"):
                    continue
                matched = True
                break
            if matched:
                hits.append(off)

        if len(hits) == 1:
            return hits[0]
        return None

File: sources/FirmwarePatcher/Kernel/JBPatches/KernelJBPatchVmFault.swift
// KernelJBPatchVmFault.swift — JB kernel patch: VM fault enter prepare bypass
//
// Historical note: derived from the legacy Python firmware patcher during the Swift migration.

import Capstone
import Foundation

extension KernelJBPatcher {
    /// Force the upstream cs_bypass fast-path in `_vm_fault_enter_prepare`.
    ///
    /// Expected semantic shape:
    ///   ... early in prologue: LDR Wflags, [fault_info_reg, #0x28]
    ///   ... later:             TBZ Wflags, #3, validation_path
    ///                          MOV Wtainted, #0
    ///                          B   post_validation_success
    ///
    /// NOPing the TBZ forces the fast-path unconditionally.
    @discardableResult
    func patchVmFaultEnterPrepare() -> Bool {
        log("\n[JB] _vm_fault_enter_prepare: NOP")

        var candidateFuncs: [Int] = []

        // Strategy 1: symbol table lookup
        if let funcOff = resolveSymbol("_vm_fault_enter_prepare") {
            candidateFuncs.append(funcOff)
        }

        // Strategy 2: string anchor
        if let strOff = buffer.findString("vm_fault_enter_prepare") {
            let refs = findStringRefs(strOff)
            for (adrpOff, _) in refs {
                if let funcOff = findFunctionStart(adrpOff) {
                    candidateFuncs.append(funcOff)
                }
            }
        }

        var candidateSites = Set<Int>()
        for funcStart in Set(candidateFuncs) {
            let funcEnd = findFuncEnd(funcStart, maxSize: 0x4000)
            if let site = findCsBypassGate(start: funcStart, end: funcEnd) {
                candidateSites.insert(site)
            }
        }

        if candidateSites.count == 1 {
            let patchOff = candidateSites.first!
            let va = fileOffsetToVA(patchOff)
            emit(patchOff, ARM64.nop,
                 patchID: "kernelcache_jb.vm_fault_enter_prepare",
                 virtualAddress: va,
                 description: "NOP [_vm_fault_enter_prepare]")
            return true
        } else if candidateSites.count > 1 {
            let list = candidateSites.sorted().map { String(format: "0x%X", $0) }.joined(separator: ", ")
            log("  [-] ambiguous vm_fault_enter_prepare candidates: \(list)")
            return false
        }

        log("  [-] patch site not found")
        return false
    }

    // MARK: - Private helpers

    /// Find the unique `tbz Wflags, #3 / mov Wt, #0 / b ...` gate inside
    /// a function, where Wflags is the register loaded from [base, #0x28]
    /// in the function prologue.
    private func findCsBypassGate(start: Int, end: Int) -> Int? {
        // Pass 1: collect registers loaded from [*, #0x28] in first 0x120 bytes
        var flagRegs = Set<UInt32>()
        let prologueEnd = min(end, start + 0x120)
        var off = start
        while off < prologueEnd {
            let insns = disasm.disassemble(in: buffer.data, at: off, count: 1)
            guard let insn = insns.first else { off += 4; continue }
            if insn.mnemonic == "ldr",
               let ops = insn.aarch64?.operands, ops.count >= 2,
               ops[0].type == AARCH64_OP_REG,
               ops[1].type == AARCH64_OP_MEM,
               ops[1].mem.base != AARCH64_REG_INVALID,
               ops[1].mem.disp == 0x28
            {
                let dstName = insn.operandString.components(separatedBy: ",").first?.trimmingCharacters(in: .whitespaces) ?? ""
                if dstName.hasPrefix("w") {
                    flagRegs.insert(ops[0].reg.rawValue)
                }
            }
            off += 4
        }

        guard !flagRegs.isEmpty else { return nil }

        // Pass 2: scan for TBZ Wflags, #3, target / MOV Wt, #0 / B target2
        var hits: [Int] = []
        let scanStart = max(start + 0x80, start)
        off = scanStart
        while off + 8 < end {
            let insns = disasm.disassemble(in: buffer.data, at: off, count: 1)
            guard let gate = insns.first else { off += 4; continue }
            guard gate.mnemonic == "tbz",
                  let gateOps = gate.aarch64?.operands, gateOps.count == 3,
                  gateOps[0].type == AARCH64_OP_REG,
                  flagRegs.contains(gateOps[0].reg.rawValue),
                  gateOps[1].type == AARCH64_OP_IMM, gateOps[1].imm == 3,
                  gateOps[2].type == AARCH64_OP_IMM
            else { off += 4; continue }

            // Check mov Wt, #0
            let movInsns = disasm.disassemble(in: buffer.data, at: off + 4, count: 1)
            guard let movInsn = movInsns.first,
                  movInsn.mnemonic == "mov",
                  let movOps = movInsn.aarch64?.operands, movOps.count == 2,
                  movOps[0].type == AARCH64_OP_REG,
                  movOps[1].type == AARCH64_OP_IMM, movOps[1].imm == 0
            else { off += 4; continue }
            let movDstName = movInsn.operandString.components(separatedBy: ",").first?.trimmingCharacters(in: .whitespaces) ?? ""
            guard movDstName.hasPrefix("w") else { off += 4; continue }

            // Check unconditional B
            let bInsns = disasm.disassemble(in: buffer.data, at: off + 8, count: 1)
            guard let bInsn = bInsns.first,
                  bInsn.mnemonic == "b",
                  let bOps = bInsn.aarch64?.operands, bOps.count == 1,
                  bOps[0].type == AARCH64_OP_IMM
            else { off += 4; continue }

            hits.append(off)
            off += 4
        }

        return hits.count == 1 ? hits[0] : nil
    }
}
